### PR TITLE
Add 7zip Types

### LINKED ISSUE #163: Support 7-Zip archives
Please add support for 7-Zip (.7z) archives. Preferably also support encrypted ones.

### PR DESCRIPTION
**Pull request by N00byKing**
Closes #163 

## Background:
I wanted an open-source File Manager. Found this one, seemed nice.
OH NOES! IT DOES NOT SUPPORT MY FAVOURITE ARCHIVE FORMAT (7z)!
Good thing I know some ʳᵉᵃᶫᶫʸ, ʳᵉᵃᶫᶫʸ ᵇᵃᵈ Java and it's open-source after all!

## Implementation:
- One added Dependency: XZ for Java. Commons-Compress requires XZ for Java for 7zip (http://commons.apache.org/proper/commons-compress/limitations.html)
- Add 7z extension to RecyclerAdapter, CompressedHelper and MimeTypes
- Implements SevenZipHelperTask, SevenZipExtractor, SevenZipDecompressor

- SevenZipHelperTask is pretty much copy-pasted from Gzip and rebranded
- SevenZipExtractor is not as good as could be, I admit. It does is job though. One problem: It doesn't buffer, as I couldn't find a way to do it with commons-compress
- SevenZipDecompressor is also copy-pasted from Gzip, rebranded

This is probably my first and last meaningful addition here, cuz I really don't know that much about file managers :P 

Sorry for it being this bad, just wanted some 7z support.

And lastly, thank you for making this app!

P.S. "7zip type**s**", as 7zip archives can use different compression algorithms

### CODE REVIEW

**Comment by EmmanuelMess** [comment]
v3.3.0 is feature complete. This is stuck until it's released.
Also, and I don't mean this as a reprimand, but this issue was assigned to me, you should have asked for progress because I would have told you about the feature completeness and that I have already done some work on it: [feature-sevenzip](https://github.com/EmmanuelMess/AmazeFileManager/commits/feature-sevenzip).

**Comment by VishalNehra** [comment]
@N00byKing nice branch name you got there

**Comment by N00byKing** [comment]
You mean 'Best Branch Name 2018', right?

**Comment by VishalNehra** [comment]
Working as expected. Thanks for your contribution (:

**Comment by N00byKing** [comment]
^^ thank you!

**Comment by VishalNehra** [comment]
@TranceLove 
1. This PR wasn't intending on adding support to view 7z files.
2. We can read/write individual int instead of reading a whole entry in byte array to avoid OOM. That'll slow down the extraction a bit but it'll also give a consistent progress.
Or we can use some other API altogether since commons compress doesn't support streaming 7z files.

**Comment by N00byKing** [comment]
Regarding viewing files: I'll look into it. While it wasn't originally planned for this PR, I don't object to adding it
I'll try to add buffering too, but lets see how that goes. As I said, some *bad* java

**Comment by VishalNehra** [comment]
@TranceLove correctly pointed, I implemented a quick fix as per your suggestion, it does increase the extraction speeds by 3-4 times. Code snipped [here](https://pastebin.com/vcxdrzvT).
I was getting 300-400KBPS while copying individual byte, now getting 0.8 - 1 MBPS 
However still nowhere close to the likes of solid file explorer which gives a consistent 4-5 MBPS under a sequential read.

File: app/build.gradle [hunk 1]
```diff
@@ -111,6 +111,8 @@ dependencies {
     implementation group: 'org.slf4j', name: 'slf4j-api', version: '1.7.25'
     // https://mvnrepository.com/artifact/org.slf4j/slf4j-simple
     implementation group: 'org.slf4j', name: 'slf4j-simple', version: '1.7.21'
+    // Dependency for 7zip Decompression
+    implementation group: 'org.tukaani', name: 'xz', version: '1.8'
```
[the remark this answers is not in the document]
Comment (N00byKing) [reply app/build.gradle]: It is a dependency of commons-compress.
(http://commons.apache.org/proper/commons-compress/limitations.html)

**Comment by TranceLove** [comment]
Tested again with my Oneplus One running Slim7 (7.1.2), now it can extract the big file without crash. Many thanks for the work!

> Pps. another few tests and speeds are really good, sometimes over 3 MBPS. I think this should be good to go unless anyone got better ways.

At my test I got around 2MBps. Actually I thought the byte buffer size can be bigger (16384 bytes? Or even 4194304 bytes = 4MB?), but personally I'd accept the current setting.

(Actually, bigger buffer size doesn't necessary means faster I/O since flushing buffer also takes time, AFAIK)

Had some ideas in mind on speeding up the I/O, from using NIO's `ByteBuffer` to a [circular buffer](https://en.wikipedia.org/wiki/Circular_buffer) which one side feeds in bytes and the other side took the bytes to a file OutputStream.

For a ring buffer, Google took me to [Greenrobot Essentials](http://greenrobot.org/essentials/), which seems promising.

One way another, I'd suggest doing such optimisation beyond the 3.3.0 milestone unless we do have plenty of time to test out the refactoring; especially if we're to use Greenrobot's library the refactoring will take a bigger scale since other parts of the app should also benefit from the library.

No problems in sight, giving a +1.

**Comment by VishalNehra** [comment]
Now that we're at it, would you be interested in adding support for viewing gzip files in a separate PR @N00byKing ?

**Comment by N00byKing** [comment]
¯\_(ツ)_/¯
I actually came for 7zip as originally stated but tbh its quite fun working on this stuff.
Will do as soon as I get back home

**Comment by VishalNehra** [comment]
That'd be greatly appreciated (:

**Comment by TranceLove** [comment]
> Now that we're at it, would you be interested in adding support for viewing gzip files in a separate PR @N00byKing ?

Viewing gzip archives should have been available since #991 is merged.

**Comment by VishalNehra** [comment]
Ahh, my bad. We don't get any prompt to open it in Amaze, and might've slipped that PR. Maybe there's no entry in intent-filter to handle it?

**Comment by TranceLove** [comment]
> Ahh, my bad. We don't get any prompt to open it in Amaze, and might've slipped that PR. Maybe there's no entry in intent-filter to handle it?

Nice catch - somebody forgot to fill it in to the manifest (as of 0dd7ed3). I'll leave this to you guys then :)

**Comment by TranceLove** [comment]
All agreed, pushing through.

**Comment by EmmanuelMess** [comment]
This should not have been merged, 3.3.0 is feature complete. Merge should be reverted. @VishalNehra ?

**Comment by TranceLove** [comment]
Oops...

**Comment by N00byKing** [comment]
Well then, looks like my time here is over.

It was nice meeting you all here, good luck with the project!

**Comment by VishalNehra** [comment]
@N00byKing thanks again for your contribution (:

**Pull request merged**